ambari-server/src/main/java/org/apache/ambari/server/upgrade/UpgradeCatalog300.java

```diff
@@ -784,9 +888,50 @@ protected void updateKerberosConfigurations() throws AmbariException {
                   true, false);
             }
           }
+          if (doPrincipalsCreation) {
+            PrepareKerberosIdentitiesServerAction prepareIdentities = new PrepareKerberosIdentitiesServerAction();
+            ExecutionCommand executionCommand = new ExecutionCommand();
+            executionCommand.setCommandParams(new HashMap<String, String>(){{
+              put(KerberosServerAction.DEFAULT_REALM, config.getProperties().get("realm"));
+            }});
+            prepareIdentities.setExecutionCommand(executionCommand);
+
+            // inject whatever we need for calling desired server action
+            injector.injectMembers(prepareIdentities);
+            KerberosHelper kerberosHelper = injector.getInstance(KerberosHelper.class);
+
+            injector.getInstance(AmbariServer.class).performStaticInjection();
+            AmbariServer.setController(injector.getInstance(AmbariManagementController.class));
+
+            KerberosDescriptor kerberosDescriptor = kerberosHelper.getKerberosDescriptor(cluster, false);
+            Map<String, Map<String, String>> kerberosConfigurations = new HashMap<>();
+            Map<String, Set<String>> propertiesToIgnore = new HashMap<>();
+            List<ServiceComponentHost> schToProcess = kerberosHelper.getServiceComponentHostsToProcess(cluster, kerberosDescriptor, null, null);
+            Map<String, Map<String, String>> configurations = kerberosHelper.calculateConfigurations(cluster, null, kerberosDescriptor, false, false);
+            boolean includeAmbariIdentity = true;
+            String dataDirectory = kerberosHelper.createTemporaryDirectory().getAbsolutePath();
+            List<Boolean> failed = new ArrayList<>();
```

Comment:
Why is this a List?

Reply:
runnable is an anonymous inner class, so failed must be final, which does not allow to modify variable in runnable. But as far as I see this is not needed at all, I will just re-raise AmbariException as RuntimeException in run method and re-wrap it in AmbariException again